### PR TITLE
Enforce pipe ownership from client side

### PR DESCRIPTION
**Pull request by zanderz**
Modifies the windows "socket" object to also apply the check to make sure the pipe is owned by the same user who started the current service.
This check fails when the pipe server and client are in the same process, so the test case is distinguished.
GUI performs check via shelling out to command line and doing `keybase pipeowner ...`
Sample client installer with all these changes, plus a throwaway KBFS branch using the new pipe restrictions, is at https://prerelease.keybase.io/windows/Keybase_1.0.48-20180511140352%2B124010c67a.386.exe

### CODE REVIEW

**Comment by maxtaco** [comment]
Needs tests in Go.

File: go/client/cmd_pipe_owner.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+func (s *CmdPipeOwner) ParseArgv(ctx *cli.Context) error {
+	if len(ctx.Args()) < 1 {
+		return errors.New("pipeowner needs a pipe name")
+	}
+	s.arg = ctx.Args()[0]
+	return nil
+}
+
+func (s *CmdPipeOwner) Run() error {
+	owner, err := libkb.Pipeowner(s.arg)
```
Comment (maxtaco) [inline go/client/cmd_pipe_owner.go]: preferred Go style is to error out ASAP.

```go
if err != nil {
    return err
}
```

File: go/client/cmd_pipe_owner.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	if err == nil {
+		dui := s.G().UI.GetDumbOutputUI()
+		dui.Printf("%v\n", owner)
+		if !owner {
+			err = errors.New("failed to establish pipe ownership")
```
Comment (maxtaco) [inline go/client/cmd_pipe_owner.go]: Preferred Go style:

```go
return err
```

**Comment by zanderz** [comment]
@maxtaco Testing the failure case requires running a process under a different windows user's terminal session, I don't see how to do that in a go test?

**Comment by maxtaco** [comment]
Yeah, I was thinking about that... Can you make throwaway users as a non-admin?  Or how about different SIDs for the current user? 

Well you can at least test some error cases, like file not found, etc.  

Also, we should probably get a review from another Windows programmer, like @taruti or @zapu

**Comment by zanderz** [comment]
Most devs and ci machines will be running as admin, and you can make test accounts no problem and run processes as another user, but IDK how to completely automate it so that no twitchy manual setup would be necessary

**Comment by maxtaco** [comment]
Is there an API for making a new user?  There is on linux, I believe...  Or you can just pick a random UID and setuid a process to it, the OS will allow it.

File: go/libkb/pipeowner_windows.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
+// Copyright 2015 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package libkb
```
Comment (zapu) [inline go/libkb/pipeowner_windows.go]: should we conditionally compile this on windows only?

also 2018 instead of 2015 in copyright notice

**Comment by zanderz** [comment]
There seems to be a way - there is a distinction between "net login" and "interactive logon": https://support.microsoft.com/en-us/help/196070/how-to-programmatically-cause-the-creation-of-a-user-s-profile
I could try calling these from Go or just making a C utility to do it, and of course run as that user to open a named pipe.

File: go/libkb/pipeowner_windows.go (hunk 3 above)
Comment (zanderz) [reply go/libkb/pipeowner_windows.go]: The `_windows` part of the filename makes it conditionally compiled, but I don't mind adding both

File: go/libkb/security_windows.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
+// copied from github.com/hectane/go-acl
+
```
Comment (taruti) [inline go/libkb/security_windows.go]: Is this enough attribution?

File: go/libkb/pipeowner_windows.go [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+	return "All pipe instances are busy"
+}
+
+func waitNamedPipe(name string, timeout uint32) (err error) {
+	rawName, e1 := syscall.UTF16PtrFromString(name)
+	if e1 != nil {
+		return e1
+	}
+
+	r1, _, e2 := syscall.Syscall(procWaitNamedPipeW.Addr(), 2, uintptr(unsafe.Pointer(rawName)), uintptr(timeout), 0)
```
Comment (taruti) [inline go/libkb/pipeowner_windows.go]: procWaitNamedPipeW.call

Comment (taruti) [inline go/libkb/pipeowner_windows.go]: r1, _, e2 := procWaitNamedPipeW.Call(...)
if r1 == 0 {
  return e2
}
return nil

File: go/libkb/security_windows.go (hunk 4 above)
Comment (maxtaco) [reply go/libkb/security_windows.go]: It's a good point, you should check if this changes our licensing at all. Which license is this code under?

File: go/libkb/pipeowner_windows.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+	var secDesc windows.Handle
+
+	err := GetNamedSecurityInfo(name, SE_FILE_OBJECT, OWNER_SECURITY_INFORMATION, &userSID, nil, nil, nil, &secDesc)
+	if err != nil {
+		return nil, err
+	}
+	return userSID, nil
+}
+
+func Pipeowner(name string) (bool, error) {
```
Comment (maxtaco) [inline go/libkb/pipeowner_windows.go]: If this is a predicate, can you call it `IsPipeowner`?

File: go/libkb/socket_windows_test.go [hunk 7]
```diff
@@ -79,3 +79,20 @@ func namedPipeClient(sendSocket Socket, t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
+func TestWindowsPipeOwnerNotexist(t *testing.T) {
```
Comment (maxtaco) [inline go/libkb/socket_windows_test.go]: Capitalize the `E` in `exist`

File: go/libkb/socket_windows_test.go [hunk 8]
```diff
@@ -79,3 +79,20 @@ func namedPipeClient(sendSocket Socket, t *testing.T) {
...
+	// Test ownership
+	owner, err := Pipeowner("name_of_nonexistent_pipe")
+	if err == nil {
+		t.Fatal(errors.New("Expected error getting owner of nonexistent pipe"))
+	}
+	if owner {
+		t.Fatal(errors.New("Expected false getting owner of nonexistent pipe"))
+	}
+	
+}
```
Comment (maxtaco) [inline go/libkb/socket_windows_test.go]: Is the success case tested? I couldn't find it.

**Comment by zanderz** [comment]
BTW I spent some time building a C example of using `LoadUserProfile()` to create a user account, and it worked, but only if I opened the CMD with elevated permissions. IOW, it still seems like special setup would be required on any machine running the tests.

**Comment by taruti** [comment]
Windows side looks now ok, thanks!

**Pull request merged**